http/v1.md

```diff
@@ -0,0 +1,488 @@
+
+#### Common error codes
+
+| Code | Reason                                  |
+| ---- | --------------------------------------- |
+| 400  | Bad Request – Invalid request format    |
+| 404  | Not found                               |
+| 429  | Too many requests - Rate limit exceeded |
+| 500  | Internal Server Error                   |
+| 501  | Not Implemented                         |
+
+#### Error reporting format
+For all **400** responses, see the [error response schema](https://github.com/0xProject/0x-monorepo/blob/development/packages/json-schemas/schemas/relayer_api_error_response_schema.ts#L1).
+
+```
+{
+    "code": 101,
+    "reason": "Validation failed",
+    "validationErrors": [
+        {
+            "field": "maker",
+            "code": 1002,
+            "reason": "Invalid address"
+        }
+    ]
+}
+```
+
+General error codes:
+
+```
+100 - Validation Failed
+101 - Malformed JSON
+102 - Order submission disabled
+103 - Throttled
+```
+
+Validation error codes:
+
+```
+1000 - Required field
+1001 - Incorrect format
+1002 - Invalid address
+1003 - Address not supported
+1004 - Value out of range
+1005 - Invalid signature or hash
+1006 - Unsupported option
+```
+
+
+### Misc.
+
+*   All requests and responses should be of **application/json** content type
+*   All token amounts are sent in amounts of the smallest level of precision (base units). (e.g if a token has 18 decimal places, selling 1 token would show up as selling `'1000000000000000000'` units by this API).
+*   All addresses are sent as lower-case (non-checksummed) Ethereum addresses with the `0x` prefix.
+
+## REST API
+
+### GET /v1/asset_pairs
+
+Retrieves a list of available asset pairs and the information required to trade them. This endpoint should be [paginated](#pagination).
+
+#### Parameters
+
+*   assetDataA=&assetDataB [string]: these are assetData fields. Returns asset pairs that contain assetDataA and assetDataB (in any order). Setting only assetDataA or assetDataB returns pairs filtered by that asset only (optional)
+
+#### Response
+
+[See response schema](https://github.com/0xProject/0x-monorepo/blob/development/packages/json-schemas/schemas/relayer_api_token_pairs_response_schema.ts#L1)
+
+```
+[
+    "total": 43,
+    "page": 1,
+    "per_page": 100,
+    "records": [
+        {
+            "assetDataA": {
+                "minAmount": "0",
+                "maxAmount": "10000000000000000000",
+                "precision": 5
+            },
+            "assetDataB": {
+                "minAmount": "0",
+                "maxAmount": "50000000000000000000",
+                "precision": 5
+            }
+        },
+        ...
+    ]
+]
+```
+
+*   `minAmount` - the minimum trade amount the relayer will accept
+*   `maxAmount` - the maximum trade amount the relayer will accept
+*   `precision` - the desired price precision a relayer would like to support within their orderbook
```

Comment:
Hm... these only seem to make sense for fungible assets... For non-fungibles, `minAmount` & `maxAmount` would always both be 1, and `precision` would always be 0.

Maybe we need something similar but slightly different for non-fungibles? I feel like people would be interested in knowing which non-fungibles are tradeable, and against what fungible tokens. This might be out-of-scope for SRA... Thoughts?

Reply:
Yeah, this was a recurring problem in discussion (things that make sense for fungibles are kind of clunky for NFTs). IMO, this isn't terrible as it's at least compatible with NFTs (using the method you described), if redundant.

Re: knowing trading pairs, isn't this accomplished by just calling asset_pairs? Are you thinking of some way to parameterize calls by AssetProxy to return e.g. all ERC-20 token pairs or all ERC-721 pairs? Sorry if I'm missing something.

Reply:
Do we want to keep this endpoint? Maybe we just have the API return an error when these requirements are not met.

Reply:
I know there are some organizations that pull liquidity by querying for token pairs, iterating over the token pairs and querying the orders / orderbook API for each token pair. Getting rid of the token pairs API completely would make it hard to adjust those implementations.